src/Npgsql.NetTopologySuite/Internal/NetTopologySuiteTypeInfoResolver.cs

```diff
@@ -0,0 +1,98 @@
+using System;
+using NetTopologySuite;
+using NetTopologySuite.Geometries;
+using NetTopologySuite.IO;
+using Npgsql.Internal;
+using Npgsql.Internal.Postgres;
+
+namespace Npgsql.NetTopologySuite.Internal;
+
+sealed class NetTopologySuiteTypeInfoResolver : IPgTypeInfoResolver
+{
+    TypeInfoMappingCollection Mappings { get; }
+
+    public NetTopologySuiteTypeInfoResolver(
+        CoordinateSequenceFactory? coordinateSequenceFactory,
+        PrecisionModel? precisionModel,
+        Ordinates handleOrdinates,
+        bool geographyAsDefault)
+    {
+        coordinateSequenceFactory ??= NtsGeometryServices.Instance.DefaultCoordinateSequenceFactory;
+        precisionModel ??= NtsGeometryServices.Instance.DefaultPrecisionModel;
+        handleOrdinates = handleOrdinates == Ordinates.None ? coordinateSequenceFactory.Ordinates : handleOrdinates;
+
+        var reader = new PostGisReader(coordinateSequenceFactory, precisionModel, handleOrdinates);
+        var writer = new PostGisWriter();
+
+        Mappings = new TypeInfoMappingCollection();
+        AddInfos(Mappings, reader, writer, geographyAsDefault);
+        // TODO: Opt-in only
+        AddArrayInfos(Mappings);
+    }
+
+    public PgTypeInfo? GetTypeInfo(Type? type, DataTypeName? dataTypeName, PgSerializerOptions options)
+        => Mappings.Find(type, dataTypeName, options);
+
+    static void AddInfos(TypeInfoMappingCollection mappings, PostGisReader reader, PostGisWriter writer, bool geographyAsDefault)
```

Comment:
When it comes to TypeMatchPredicate or not; There's two thoughts: the approach here is purely data driven so if we do decide to add lookups it's the fastest, if we don't, having less mappings to scan can help.

Reply:
Sure, TypeMatchPredicate does sound like a good way to do it - frankly I'm not sure the perf (less mappings) matters, but if I'm understanding correctly it will also simplify the resolver quite significantly, which is nice.

Reply:
Am happy with whatever works!

Reply:
Right, I forgot, it'd need activator.createinstance for the T. (we can suppress safely but let's leave it like this...)